### PR TITLE
thors-serializer: fix build for Linux

### PR DESCRIPTION
**Pull request by cho-m**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

https://github.com/Homebrew/homebrew-core/runs/3096318546?check_suite_focus=true
```
checking whether g++-5 accepts -g... yes
checking for python... no
configure: error: Cannot find python in your system path
```

### CODE REVIEW

**Comment by danielnachun** [comment]
I was able to get further into the build process by adding
```
   ENV["PYTHON"] = "#{Formula["python@3.9"].opt_bin}/python3.9"
```
We also need to use a newer GCC, so I had to add
```
on_linux do
   depends_on "gcc"
end
fails_with gcc: "5"
```
The error I got stuck at was 
```
In file included from FrameTest.cpp:51:
testlist.h:7:13: error: extended character © is not valid in an identifier
    7 | # Copyright © 2013 Bart Massey
      |             ^
```
which is a trivial problem, but my attempts to fix it have not been a success so far.

**Comment by cho-m** [comment]
During build-time tests, it looks like coverage test isn't adding `BsonPrinter`. Could give LLVM a try instead.
```make
g++-11 -o coverage/unittest.app -DCOVERAGE_test -DTHOR_COVERAGE      ERROR
export RPATH=/usr/local/lib;
g++-11 -o coverage/unittest.app coverage/SerSetTest.o coverage/PolyMorphicSerializerUniquePointerTest.o coverage/YamlPrinterTest.o coverage/DeprecatedTest.o coverage/TemplateTypeTest.o coverage/BankAccountTest.o coverage/SerDequeTest.o coverage/IgnoreUneededDataTest.o coverage/OverrideIdentiferToKeyMapTest.o coverage/SerVectorTest.o coverage/YamlParserTest.o coverage/SerMapTest.o coverage/SerInitializerListTest.o coverage/SerMemoryTest.o coverage/CornerCaseTest.o coverage/Issue72Test.o coverage/unittest.o coverage/SerMultiMapTest.o coverage/SerUnorderedMultiSetTest.o coverage/Issue42Test.o coverage/SerTuppleTest.o coverage/BsonParserTest.o coverage/SerUnorderedSetTest.o coverage/HomeBrewTest.o coverage/Issue49Test.o coverage/EscapeControlTest.o coverage/RoundTripTest.o coverage/StaticMemberTest.o coverage/SerializeEnumTest.o coverage/Issue50Test.o coverage/LexerTest.o coverage/SerUnorderedMapTest.o coverage/SerListTest.o coverage/Logging.o coverage/ReferenceWrapperTest.o coverage/SerializeTest.o coverage/PolymorphicTest.o coverage/BsonUtilitySerializationTest.o coverage/ParserInterfaceTest.o coverage/BsonLongArrayTest.o coverage/SmartPointerTest.o coverage/PointerTest.o coverage/BsonPrinterTest.o coverage/JsonPrinterTest.o coverage/SerUnorderedMultiMapTest.o coverage/SerMultiSetTest.o coverage/Issue38Test.o coverage/JsonParserTest.o coverage/TwitterTest.o coverage/UnicodeIteratorTest.o coverage/Issue51Test.o coverage/SerArrayTest.o coverage/ExceptionWhilePrintingTest.o coverage/FilterValuesTest.o coverage/unittest.o -L./coverage -lUnitTest -L../coverage -L/tmp/thors-serializer-20210719-9466-1qqwrfl/build/dlib -L/tmp/thors-serializer-20210719-9466-1qqwrfl/build/lib -lgtest -fprofile-arcs -ftest-coverage -lpthread -L/tmp/thors-serializer-20210719-9466-1qqwrfl/build/lib -lThorsLogging17D -L/usr/local/lib -lyaml -ldl
===================================================
/home/linuxbrew/.linuxbrew/bin/ld: coverage/PolyMorphicSerializerUniquePointerTest.o: in function `ThorsAnvil::Serialize::SerializeMember<PolyMorphicSerializerUniquePointerTest::AbstractTourResult, std::unique_ptr<PolyMorphicSerializerUniquePointerTest::Quantities, std::default_delete<PolyMorphicSerializerUniquePointerTest::Quantities> >, ThorsAnvil::Serialize::Traits<std::remove_cv<std::unique_ptr<PolyMorphicSerializerUniquePointerTest::Quantities, std::default_delete<PolyMorphicSerializerUniquePointerTest::Quantities> > >::type>::type> ThorsAnvil::Serialize::make_SerializeMember<PolyMorphicSerializerUniquePointerTest::AbstractTourResult, std::unique_ptr<PolyMorphicSerializerUniquePointerTest::Quantities, std::default_delete<PolyMorphicSerializerUniquePointerTest::Quantities> > >(ThorsAnvil::Serialize::Serializer&, ThorsAnvil::Serialize::PrinterInterface&, PolyMorphicSerializerUniquePointerTest::AbstractTourResult const&, std::pair<char const*, std::unique_ptr<PolyMorphicSerializerUniquePointerTest::Quantities, std::default_delete<PolyMorphicSerializerUniquePointerTest::Quantities> > PolyMorphicSerializerUniquePointerTest::AbstractTourResult::*> const&) [clone .isra.0]':
PolyMorphicSerializerUniquePointerTest.cpp:(.text+0xa3c): undefined reference to `void ThorsAnvil::Serialize::BsonPrinter::writeInt<4ul, int>(int)'
/home/linuxbrew/.linuxbrew/bin/ld: coverage/PolyMorphicSerializerUniquePointerTest.o: in function `ThorsAnvil::Serialize::BsonPrinter::addValue(int)':
PolyMorphicSerializerUniquePointerTest.cpp:(.text._ZN10ThorsAnvil9Serialize11BsonPrinter8addValueEi[_ZN10ThorsAnvil9Serialize11BsonPrinter8addValueEi]+0xd): undefined reference to `void ThorsAnvil::Serialize::BsonPrinter::writeInt<4ul, int>(int)'
/home/linuxbrew/.linuxbrew/bin/ld: coverage/Issue42Test.o: in function `ThorsAnvil::Serialize::BsonPrinter::addValue(double)':
Issue42Test.cpp:(.text._ZN10ThorsAnvil9Serialize11BsonPrinter8addValueEd[_ZN10ThorsAnvil9Serialize11BsonPrinter8addValueEd]+0xd): undefined reference to `void ThorsAnvil::Serialize::BsonPrinter::writeFloat<8ul, double>(double)'
/home/linuxbrew/.linuxbrew/bin/ld: coverage/Issue42Test.o: in function `void ThorsAnvil::Serialize::printPolyMorphicObject<Issue42Test::Vehicle>(ThorsAnvil::Serialize::Serializer&, ThorsAnvil::Serialize::PrinterInterface&, Issue42Test::Vehicle const&)':
Issue42Test.cpp:(.text._ZN10ThorsAnvil9Serialize22printPolyMorphicObjectIN11Issue42Test7VehicleEEEvRNS0_10SerializerERNS0_16PrinterInterfaceERKT_[_ZN10ThorsAnvil9Serialize22printPolyMorphicObjectIN11Issue42Test7VehicleEEEvRNS0_10SerializerERNS0_16PrinterInterfaceERKT_]+0x1e1): undefined reference to `void ThorsAnvil::Serialize::BsonPrinter::writeInt<4ul, int>(int)'
/home/linuxbrew/.linuxbrew/bin/ld: coverage/BsonPrinterTest.o: in function `BsonPrinterTest_ArrayWithArrayValues_Test::TestBody()':
BsonPrinterTest.cpp:(.text+0x25a8): undefined reference to `void ThorsAnvil::Serialize::BsonPrinter::writeInt<4ul, int>(int)'
/home/linuxbrew/.linuxbrew/bin/ld: BsonPrinterTest.cpp:(.text+0x25d5): undefined reference to `void ThorsAnvil::Serialize::BsonPrinter::writeFloat<8ul, double>(double)'
/home/linuxbrew/.linuxbrew/bin/ld: BsonPrinterTest.cpp:(.text+0x25ea): undefined reference to `void ThorsAnvil::Serialize::BsonPrinter::writeInt<8ul, long>(long)'
/home/linuxbrew/.linuxbrew/bin/ld: BsonPrinterTest.cpp:(.text+0x25ff): undefined reference to `void ThorsAnvil::Serialize::BsonPrinter::writeInt<8ul, long long>(long long)'
```

File: Formula/thors-serializer.rb [hunk 1]
```diff
@@ -16,11 +16,27 @@ class ThorsSerializer < Formula
...
+
+  fails_with gcc: 5
+
   def install
     ENV["COV"] = "gcov"
+    on_linux { ENV["PYTHON"] = Formula["python@3.9"].opt_bin/"python3" }
+
+    # Workaround for GCC error: extended character is not valid in an identifier
+    inreplace "third/ThorsIOUtil/src/ThorsIOUtil/test/testlist.h",
+              /# Copyright.*?(\d{4})/, "# Copyright \\1"
```
Comment (SMillerDev) [inline Formula/thors-serializer.rb:34]: Can this be submitted upstream?

**Comment by danielnachun** [comment]
I tried to get this to build with LLVM but it did not work either.  I think we may need to open an upstream issue about this (along with reporting the issue with invalid characters as suggested by @SMillerDev).

**Comment by cho-m** [comment]
I opened an upstream issue https://github.com/Loki-Astari/ThorsSerializer/issues/73 to hopefully get some advice on what is issue.

The only other thing I can try is stepping through older GCC/LLVM versions to see if one works.

**Comment by cho-m** [comment]
Upstream owner is working on this in #82845.

I will probably close this PR once some headway is made over there.

**Comment by cho-m** [comment]
Close in favor of #82845, which now successfully builds on Linux.

**Pull request closed without merging**